### PR TITLE
Feat: bypass router for minting CL positions

### PR DESCRIPTION
**Pull request by gasperbr**

### CODE REVIEW

File: contracts/pool/concentrated/ConcentratedLiquidityPool.sol [hunk 1]
```diff
@@ -154,46 +146,48 @@ contract ConcentratedLiquidityPool is IPool {
         }
     }
 
-    /// @dev Mints LP tokens - should be called via the router after transferring `bento` tokens.
-    /// The router must ensure that sufficient liquidity has been minted.
-    function mint(bytes calldata data) public override lock returns (uint256 _liquidity) {
+    /// @dev Mints LP tokens - should be called via the CL pool manager contract.
+    function mint(bytes calldata data) public lock returns (uint256 liquidityMinted) {
         MintParams memory mintParams = abi.decode(data, (MintParams));
```
Comment (maxsam4) [inline contracts/pool/concentrated/ConcentratedLiquidityPool.sol:151]: Don't need the params to be encoded anymore

File: contracts/pool/concentrated/ConcentratedLiquidityPool.sol [hunk 2]
```diff
@@ -154,46 +146,48 @@ contract ConcentratedLiquidityPool is IPool {
...
             priceUpper,
             currentPrice,
-            mintParams.amount1Desired,
-            mintParams.amount0Desired
+            uint256(mintParams.amount1Desired),
+            uint256(mintParams.amount0Desired)
         );
 
+        // Ensure no overflow happens when we cast from uint256 to int128.
+        if (liquidityMinted > uint128(type(int128).max)) revert Overflow();
```
Comment (maxsam4) [inline contracts/pool/concentrated/ConcentratedLiquidityPool.sol:168]: Unless required by the compiler, typecasting to uint128 is not required. int128 is smaller.

File: contracts/pool/concentrated/ConcentratedLiquidityPool.sol [hunk 3]
```diff
@@ -212,16 +206,11 @@ contract ConcentratedLiquidityPool is IPool {
...
 
-        {
-            ITridentRouter.TokenInput[] memory callbackData = new ITridentRouter.TokenInput[](2);
-            callbackData[0] = ITridentRouter.TokenInput(token0, mintParams.token0native, amount0Actual);
-            callbackData[1] = ITridentRouter.TokenInput(token1, mintParams.token1native, amount1Actual);
-            ITridentCallee(msg.sender).tridentMintCallback(abi.encode(callbackData));
-        }
+        IPositionManager(msg.sender).mintCallback(token0, token1, amount0Actual, amount1Actual, mintParams.native);
 
         unchecked {
```
Comment (maxsam4) [inline contracts/pool/concentrated/ConcentratedLiquidityPool.sol:215]: don't we need safemath to do `reserve0 += amount0Actual;` ?

File: contracts/pool/concentrated/ConcentratedLiquidityPool.sol [hunk 4]
```diff
@@ -235,52 +224,32 @@ contract ConcentratedLiquidityPool is IPool {
...
+    function burn(
         int24 lower,
         int24 upper,
-        uint128 amount,
-        address recipient,
-        bool unwrapBento
+        uint128 amount
     )
         public
         returns (
```
Comment (maxsam4) [inline contracts/pool/concentrated/ConcentratedLiquidityPool.sol:236]: Slap `lock` for good measure?

File: contracts/pool/concentrated/ConcentratedLiquidityPool.sol [hunk 5]
```diff
@@ -289,56 +258,32 @@ contract ConcentratedLiquidityPool is IPool {
...
-
-            uint256 amount0fees;
-            uint256 amount1fees;
-            (amount0fees, amount1fees, oldLiquidity) = _updatePosition(msg.sender, lower, upper, -int128(amount));
-
-            withdrawnAmounts = new TokenAmount[](2);
-            withdrawnAmounts[0] = TokenAmount({token: token0, amount: amount0});
-            withdrawnAmounts[1] = TokenAmount({token: token1, amount: amount1});
+        // Ensure no overflow happens when we cast from uint128 to int128.
+        if (amount > uint128(type(int128).max)) revert Overflow();
```
Comment (maxsam4) [inline contracts/pool/concentrated/ConcentratedLiquidityPool.sol:262]: If the compiler allows...
```suggestion
        if (amount > type(int128).max) revert Overflow();
```

File: contracts/pool/concentrated/ConcentratedLiquidityPool.sol [hunk 6]
```diff
@@ -289,56 +258,32 @@ contract ConcentratedLiquidityPool is IPool {
...
 
         unchecked {
+            amount0 = token0Amount + token0Fees;
+            amount1 = token1Amount + token1Fees;
             reserve0 -= uint128(amount0);
             reserve1 -= uint128(amount1);
         }
 
-        _transferBothTokens(recipient, amount0, amount1, unwrapBento);
+        _transferBothTokens(msg.sender, amount0, amount1, false);
```
Comment (maxsam4) [inline contracts/pool/concentrated/ConcentratedLiquidityPool.sol:276]: The `_transferBothTokens` function doesn't need to handle unwrapBento case anymore

File: contracts/pool/concentrated/ConcentratedLiquidityPool.sol [hunk 7]
```diff
@@ -235,52 +224,32 @@ contract ConcentratedLiquidityPool is IPool {
...
+            uint256 token0Amount,
+            uint256 token1Amount,
+            uint256 token0Fees,
+            uint256 token1Fees
         )
     {
-        uint256 amount0;
-        uint256 amount1;
-
         {
```
Comment (maxsam4) [inline contracts/pool/concentrated/ConcentratedLiquidityPool.sol:243]: Wonder if we can get rid of this scope now

File: contracts/pool/concentrated/ConcentratedLiquidityPool.sol (hunk 2 above)
Comment (gasperbr) [reply contracts/pool/concentrated/ConcentratedLiquidityPool.sol:168]: It's required by the compiler since `type(int128).max`  is of type int128

**Review by sarangparikh22 — APPROVED** [review]

**Pull request merged**